lib/semantics/resolve-labels.cc

```diff
@@ -0,0 +1,832 @@
+  bool Pre(const parser::SelectTypeConstruct::TypeCase&) { return SwScope(); }
+  bool Pre(const parser::WhereConstruct& W) { return PushNonBlockName(W); }
+  bool Pre(const parser::ForallConstruct& F) { return PushNonBlockName(F); }
+
+  void Post(const parser::ProgramUnit&) { PopScope(); }
+  void Post(const parser::AssociateConstruct& A) { PopName(A); }
+  void Post(const parser::BlockConstruct& Blk) { PopName(Blk); }
+  void Post(const parser::ChangeTeamConstruct& Ctm) { PopName(Ctm); }
+  void Post(const parser::CriticalConstruct& Crit) { PopName(Crit); }
+  void Post(const parser::DoConstruct& Do) { PopName(Do); }
+  void Post(const parser::IfConstruct& If) { PopName(If); }
+  void Post(const parser::CaseConstruct& Case) { PopName(Case); }
+  void Post(const parser::SelectRankConstruct& SelRk) { PopName(SelRk); }
+  void Post(const parser::SelectTypeConstruct& SelTy) { PopName(SelTy); }
+
+  //  Named constructs without block scope
+  void Post(const parser::WhereConstruct& W) { PopNonBlockConstructName(W); }
+  void Post(const parser::ForallConstruct& F) { PopNonBlockConstructName(F); }
+
+  //  Statements with label references
+  void Post(const parser::LabelDoStmt& Do) { AddDoBase(std::get<1>(Do.t)); }
+  void Post(const parser::GotoStmt& Goto) { AddBase(Goto.v); }
+  void Post(const parser::ComputedGotoStmt& C) { AddBase(std::get<0>(C.t)); }
+  void Post(const parser::ArithmeticIfStmt& AIf) {
+    AddBase(std::get<1>(AIf.t));
+    AddBase(std::get<2>(AIf.t));
+    AddBase(std::get<3>(AIf.t));
+  }
+  void Post(const parser::AssignStmt& Assn) { AddBase(std::get<0>(Assn.t)); }
+  void Post(const parser::AssignedGotoStmt& A) { AddBase(std::get<1>(A.t)); }
+  void Post(const parser::AltReturnSpec& ARS) { AddBase(ARS.v); }
+
+  void Post(const parser::ErrLabel& Err) { AddBase(Err.v); }
+  void Post(const parser::EndLabel& End) { AddBase(End.v); }
+  void Post(const parser::EorLabel& Eor) { AddBase(Eor.v); }
+  void Post(const parser::Format& Fmt) {
+    // BUG: the label is saved as an IntLiteralConstant rather than a Label
+#if 0
+    if (const auto* P{std::get_if<parser::Label>(&Fmt.u)})
+      AddFmtBase(*P);
+#else
+    // FIXME: this is wrong, but extracts the label's value
+    if (const auto* P{std::get_if<0>(&Fmt.u)}) {
+      parser::Label L{std::get<0>(std::get<parser::IntLiteralConstant>(std::get<parser::LiteralConstant>((*P->thing).u).u).t)};
+      AddFmtBase(L);
+    }
+#endif
+  }
+  void Post(const parser::CycleStmt& Cycle) {
+    if (Cycle.v.has_value())
+      CheckLabelContext("CYCLE", Cycle.v.value().ToString());
+  }
+  void Post(const parser::ExitStmt& Exit) {
+    if (Exit.v.has_value())
+      CheckLabelContext("EXIT", Exit.v.value().ToString());
+  }
+
+  // Getters for the results
+  const std::vector<UnitAnalysis>& GetProgramUnits() const { return PUnits; }
+  ErrorHandler& GetEH() { return EH; }
+  bool HasNoErrors() const { return NoErrors; }
```

Comment:
Can you implement this with `EH.messages.empty()` ? Then you wouldn't have to worry about setting clearing `NoErrors` whenever an error is reported.

Reply:
Not all the messages are errors.

Reply:
`Messages` also has `AnyFatalError()` which might match what you need.

Reply:
OK